corporate/views/remote_upgrade_page.py

```diff
@@ -0,0 +1,57 @@
+import logging
+
+from django.http import HttpRequest, HttpResponse, HttpResponseRedirect
+from django.shortcuts import render
+from django.utils.translation import gettext as _
+
+from corporate.lib.remote_billing_util import get_identity_dict_from_session
+from corporate.lib.stripe import InitialUpgradeRequest, RemoteRealmBillingSession
+from corporate.models import CustomerPlan
+from zerver.decorator import self_hosting_management_endpoint
+from zerver.lib.exceptions import JsonableError
+from zerver.lib.request import REQ
+from zerver.lib.validator import check_bool
+from zilencer.models import RemoteRealm, RemoteZulipServer
+
+billing_logger = logging.getLogger("corporate.stripe")
+
+
+@self_hosting_management_endpoint
+def remote_realm_upgrade_page(
+    request: HttpRequest,
+    *,
+    realm_uuid: str,
+    manual_license_management: bool = REQ(default=False, json_validator=check_bool),
+) -> HttpResponse:
+    identity_dict = get_identity_dict_from_session(request, realm_uuid=realm_uuid)
+
+    if identity_dict is None:
+        raise JsonableError(_("User not authenticated"))
+
+    remote_server_uuid = identity_dict["remote_server_uuid"]
+    remote_realm_uuid = identity_dict["remote_realm_uuid"]
+
+    try:
+        remote_server = RemoteZulipServer.objects.get(uuid=remote_server_uuid)
+    except RemoteZulipServer.DoesNotExist:
+        raise JsonableError(_("Invalid remote server."))
+
+    try:
+        remote_realm = RemoteRealm.objects.get(uuid=remote_realm_uuid, server=remote_server)
+    except RemoteRealm.DoesNotExist:
+        raise AssertionError(
+            "The remote realm is missing despite being in the RemoteBillingIdentityDict"
+        )
```

Comment:
I think this first part where we get into a state where we have a `RemoteRealm` object should be moved into `
get_identity_dict_from_session` (or maybe some small wrapper), since it'll be common code for all these views.

You can probably get away with a single database query with

`remote_realm = RemoteRealm.objects.get(uuid=remote_realm_uuid, server__uuid=remote_server_uuid)`.

The function can also check `remote_realm.deactivated` and `remote_realm.remote_server.deactivated` and raise a `JsonableError(_("Registration is deactivated"))` if either of those is `False`.

Reply:
Actually, it's possible what we want is to put those logic into the `self_hosting_management_endpoint` -- we might need to split that decorator into two methods: `remote_realm_management_endpoint` and `remote_server_management_endpoint`. But then we can have this function just be passed a `request: HttpRequest, remote_realm: RemoteRealm, ...`

Reply:
The decorator refactor seems like a nice change. I can do that in a follow up PR.

Reply:
Sounds good.

Reply:
Opened #27894